url04f7fcf4e1_peteratt_HDFS_Replication_Simulator_tgz-pJ8-hdfs_replicationsimulator_MainJ8/src/hdfs/replicationsimulator/AllDatanode.java

```diff
@@ -0,0 +1,261 @@
+		commandsthread.start();
+	}
+
+	/**
+	 * Stops the datanodes management thread
+	 */
+	public void stop() {
+		isRunning = false;
+
+		try {
+			hbthread.join();
+			commandsthread.join();
+
+		} catch (InterruptedException e) {
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * Send the heartbeats.
+	 */
+	class HeartbeatSender implements Runnable {
+
+		public void run() {
+			while (isRunning) {
+				try {
+					checkHearbeat();
+					Thread.sleep(heartbeatRecheckInterval);
+				} catch (Exception e) {
+					// FSNamesystem.LOG.error(StringUtils.stringifyException(e));
+				}
+			}
+		}
+
+		private void checkHearbeat() {
+			Datanode current;
+			for (int i = 0; i < datanodes.size(); i++) {
+				current = datanodes.get(i);
+				if (!current.hasFailed()
+						&& (Node.now() >= heartbeatInterval
+								+ current.getLastHB())) {
+					sendHeartbeat(i);
+					// System.out.print("HB from " + current.getId() + "\n");
+					current.setLastHB(Node.now());
+				}
+			}
+		}
+	}
+
+	public boolean sendHeartbeat(int id) {
+		return addToNamenodeQueue(new Event(id, Event.HEARTBEAT, Node.now()));
+	}
+
+	/**
+	 * Process the commands received from NameNode
+	 */
+	class CommandsHandler implements Runnable {
+
+		public void run() {
+			while (isRunning) {
+				try {
+					handleCommands();
+					Thread.sleep(communicationRecheckInterval);
+				} catch (Exception e) {
+					e.printStackTrace();
+				}
+			}
+		}
+
+		private void handleCommands() {
+			while(!toDatanodes.isEmpty()){
+				Event command = toDatanodes.poll();
+	
+				if (command != null) {
+					// Handle it
+					if (command.getAction() == Event.REPLICATION) {
+						if (!handleTransfer(command.getSource(),
+								command.getDestination(), command.getBlockId())) {
+							toDatanodes.add(command);
+						}
+					} else if(command.getAction() == Event.PENDINGTRANSFER){
+						if (!handleReception(command.getSource(),
+								command.getDestination(), command.getBlockId())) {
+							toDatanodes.add(command);
+						}
+					}
+				}
+			}
+		}
+
+		private boolean handleReception(int idSource, int idDestination, int blockId) {
+			Datanode sourceNode = null;
+			Datanode destinationNode = null;
+
+			for (Datanode node : datanodes) {
+				if (node.getId() == idSource) {
+					sourceNode = node;
+				} else if (node.getId() == idDestination) {
+					destinationNode = node;
+				}
+			}
+			boolean NodesAlive = (!sourceNode.hasFailed() && !destinationNode.hasFailed());
+			
+			if(Node.now() >= destinationNode.getDownloadingTime() && NodesAlive){
+				//Send signal to Namenode that the transfer is finished
+				toNamenode.add(new Event(destinationNode.getId(), Event.BLOCKRECEPTION, Node.now(),0,blockId));
+				return true;
+			} else if(!NodesAlive){
+				sourceNode.setUploadingTime(Node.now());
+				destinationNode.setDownloadingTime(Node.now());
+				return true;
+			} else{
+				return false;
+			}
+		}
+
+		public boolean handleTransfer(int idSource, int idDestination,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.